### PR TITLE
resource/aws_route_table: Fix route table import to avoid deleting routes

### LINKED ISSUE #2270: Updating network_interface_id in aws_route fails
### Terraform Version
Terraform v0.10.8
provider.aws 1.2.0

### Affected Resource(s)
- aws_route

### Terraform Configuration Files
Config 1:
```hcl
provider "aws" {
  version             = "~>1.2.0"
}

resource "aws_route" "route_to_eni" {
  route_table_id         = "rtb-ba9acbbb"
  destination_cidr_block = "1.2.3.0/24"

  network_interface_id = "eni-35cadddd"
}
```
Config 2 (updated network_interface_id):
```hcl
provider "aws" {
  version             = "~>1.2.0"
}

resource "aws_route" "route_to_eni" {
  route_table_id         = "rtb-ba9acbbb"
  destination_cidr_block = "1.2.3.0/24"

  network_interface_id = "eni-35cadddd"
}
```

### Debug Output
https://gist.github.com/lethalpaga/28714fe7180b2debce09489c77ba1556

### Expected Behavior
The route should have been updated with the new ENI id

### Actual Behavior
The apply fails with
```
* aws_route.route_to_eni: InvalidInstanceID: There are multiple interfaces attached to instance 'i-0ec77fe2cdde1'. Please specify an interface ID for the operation instead.
	status code: 400, request id: 060b4661-c5e8-4254-92c0-ac1d13f21df7
```

### Steps to Reproduce

1. `terraform apply` with config1 to create the route
2. `terraform apply` with config2 to attempt updating it

### Important Factoids

* The route creation works fine. Updating it fails.
* In both config1 and config2 the ENI is attached to an instance that has 2 network interfaces attached.

### References
- Possibly related to GH-1426

### LINKED ISSUE #5631: Actual routes deleted by importing aws_route_table followed by apply
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**_Importing_ an existing route table and then _applying_ causes all routes to be deleted.**

~~I'm interested in working on the solution and am preparing a PR.~~ PR #5657 fixes this issue.

The same thing happens with rules in security groups and possibly other resources. To keep this bug report focused, I'm only including the resource I'm currently attempting to fix.

### Terraform Version

Terraform v0.11.8
~~provider.aws v1.32.0~~ provider.aws v1.33.0 (verified the issue remains)

### Affected Resource(s)

* aws_route_table

### Terraform Configuration Files

This bug affects routes whether they are created in-line in the route table or separately.

```hcl
resource "aws_route_table" "route_table1" {
  vpc_id = "${var.vpc_id}"

  route {
    cidr_block = "${var.destination_cidr_ipv4}"
    vpc_peering_connection_id = "${var.vpc_peering_conn_1}"
  }
}

resource "aws_route_table" "route_table2" {
  vpc_id = "${var.vpc_id}"
}

resource "aws_route" "route1" {
  route_table_id = "${aws_route_table.route_table2.id}"
  destination_cidr_block = "${var.destination_cidr_ipv4}"
  vpc_peering_connection_id = "${var.vpc_peering_conn_2}"
}
```

### Debug Output

[Debug](https://gist.github.com/YakDriver/b9f9eb93f555339d5e2e3e4957ccfa36).

### Expected Behavior

The route tables should be left without changes.

### Actual Behavior

Both actual routes in the existing route tables are deleted.

### Steps to Reproduce

Assuming that `aws_route_table.route_table1`, `aws_route_table.route_table2`, and `aws_route.route1` already exist, both route tables will be left without routes.

```console
$ terraform import -input=false aws_route_table.route_table1 "r-rtb-07ef0b555fc5af8243887186672"
$ terraform import -input=false aws_route_table.route_table2 "r-rtb-0f27bb31b59b926bf3887186672"
$ terraform plan -input=false -out=newplan
$ terraform apply -input=false newplan
```

### References

* #704
* #3422 (possibly)
* #1426 (possibly)
* #338 (possibly)
* #225 (possibly)

### LINKED ISSUE #5686: Existing AWS routing table configuration is always being registered as a change
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.8
provider.aws v1.33.0
```

### Affected Resource(s)

* aws_route_table

### Terraform Configuration Files

```hcl
resource "aws_route_table" "prod_private" {
  vpc_id = "${aws_vpc.ivx.id}"
  count  = 3

  route {
    cidr_block     = "0.0.0.0/0"
    nat_gateway_id = "${element(aws_nat_gateway.prod_gw.*.id, count.index)}"
  }

  route {
    cidr_block = "${var.site1_local_cidr}"
    gateway_id = "${aws_vpn_gateway.vpn_gateway.id}"
  }

  route {
    cidr_block = "${var.site2_local_cidr}"
    gateway_id = "${aws_vpn_gateway.vpn_gateway.id}"
  }

  route {
    cidr_block = "${var.site3_local_cidr}"
    gateway_id = "${aws_vpn_gateway.vpn_gateway.id}"
  }

  route {
    cidr_block = "${data.terraform_remote_state.us-tf.ivx_vpc_cidr}"
    gateway_id = "${aws_vpc_peering_connection.us_west_1.id}"
  }

  tags {
    Name = "prod_private_${count.index}"
  }
}
```

### Panic Output

No panic occured.

### Expected Behavior
Once the changes have been applied by Terraform, creating a new plan shouldn't show any changes to the state.

### Actual Behavior
Everytime a plan is being created, the routing table is being modified, even though no new changes have been made to underlying terraform resources.

### Steps to Reproduce

Have an existing AWS VPC Routing Table. Modify the table to resemble the changes from the pasted hcl-snippet.
1. `terraform plan -out prod_routing.plan`
2. `terraform apply -f prod_routing.plan`
3. Create a new plan: `terraform plan -out prod_routing.plan`

### PR DESCRIPTION
**Pull request by YakDriver**
Fixes #5631, fixes #2270, fixes #5686 

Changes proposed in this pull request:

* [BUG FIX] Make importing route tables consistent with reading route tables so that imports match existing existing resources (which avoids deletion of existing routes on apply)
* Update acceptance tests:
  * Test for deleting routes bug
  * Update existing acceptance tests for the improved consistency
  * Run in AWS accounts with fewer network conflicts

Output from acceptance testing:

```console
$ make testacc TESTARGS='-run=TestAccAWSRouteTable_import'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSRouteTable_import -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSRouteTable_importBasic
--- PASS: TestAccAWSRouteTable_importBasic (50.20s)
=== RUN   TestAccAWSRouteTable_importWithCreate
--- PASS: TestAccAWSRouteTable_importWithCreate (62.47s)
=== RUN   TestAccAWSRouteTable_importComplex
--- PASS: TestAccAWSRouteTable_importComplex (188.63s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	301.337s
```

### CODE REVIEW

**Comment by bflad** [comment]
Marking as breaking change as this fundamentally changes the behavior of import for the resource.

I personally do not know the historical context of why this resource is setup to perform the complex import, but if I had to guess off top of my head, its because there is currently no method to import the separate `aws_route` resources.

**Comment by YakDriver** [comment]
Add the ability to import routes. Here are the new acceptance tests directed specifically at aws_route import:

```console
$ make testacc TESTARGS='-run=TestAccAWSRoute_import'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSRoute_import -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSRoute_importBasic
--- PASS: TestAccAWSRoute_importBasic (51.14s)
=== RUN   TestAccAWSRoute_importIPv6IGW
--- PASS: TestAccAWSRoute_importIPv6IGW (51.20s)
=== RUN   TestAccAWSRoute_importIPv6NetworkInterface
--- PASS: TestAccAWSRoute_importIPv6NetworkInterface (177.98s)
=== RUN   TestAccAWSRoute_importWithMultipleRTs
--- PASS: TestAccAWSRoute_importWithMultipleRTs (47.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	327.456s
```

File: aws/resource_aws_route.go [hunk 1]
```diff
@@ -526,3 +541,72 @@ func findResourceRoute(conn *ec2.EC2, rtbid string, cidr string, ipv6cidr string
...
+			return results, nil
+		}
+	}
+
+	return nil, err
+
+}
+
+// AWS does not provide an id for routes in route tables. To import, use a pseudo
+// ID of r-ROUTETABLEID-DESTINATION (e.g., r-rtb-12345678_10.2.0.0/16). This
```
Comment (lorengordon) [inline aws/resource_aws_route.go]: _ not - as a separator, no? i.e. `r-ROUTETABLEID_DESTINATION` not `r-ROUTETABLEID-DESTINATION`....

**Comment by YakDriver** [comment]
@bflad As you mentioned earlier, this change might be wise to consider at the same time as implementing importing `aws_route` resources. This PR now implements that enhancement.

This PR _does_ fundamentally change the import of `aws_route_table`. However, I disagree that it is a breaking change.  Here is why.

Currently, if you create an `aws_route_table` with an inline route, you'll end up with a state having an `aws_route_table` and an inline route. We'd agree that you can manage your `aws_route_table` from Terraform in that situation. It's not an anomalous state. I'd guess it's a situation a good chunk of, if not most, users find themselves in. I wouldn't say they are in a "breaking" condition.

If down the road, you import that exact same `aws_route_table`, you'll end up with a state having an `aws_route_table` and an `aws_route`. You might expect (as we did) to have the same state after the import as you did after create but, alas, you do not. The [route table documentation](https://www.terraform.io/docs/providers/aws/r/route_table.html) specifically says that `aws_route_table` and `aws_route` are not compatible resources and yet the import forces you into an anomalous state into which you have incompatible resources in your state. As a result, your routes get deleted.

This has production impact on us. When managing tenant infrastructures, when we import route tables as part of a larger import, then make changes to unrelated resources, and attempt to apply the changes, all the routes are deleted. That is the **breaking** issue. :)

**Comment by bflad** [comment]
@YakDriver I do really want to emphasize that this change does seem like a good path forward, its just a matter of keeping our compatibility promises. The `aws_route_table` import behavior has been like this for a very long time though and people may be depending on its current behavior. The maintainers are not likely to merge a change in this type of behavior before a major release, which is coming in the next few weeks to coincide with the Terraform 0.12 beta.

You have a few options for working with this today:
* Continue with the existing code and use the following workaround: `terraform state rm` the newly added `aws_route` resources found via `terraform plan` if you would like to only manage via `aws_route_table`
* Build and run a custom provider binary with the code changes contained in this pull request

My recommendations for this pull request in its current state to move things forward (unless we want to do all this in 2.0.0):
* Split the two code changes here:
  * resource/aws_route: Support resource import (this can be reviewed, adjusted, and merged today)
  * resource/aws_route_table: Remove `aws_route` resource import (this can be reviewed, adjusted and merged in 2.0.0)
* Create a separate PR to document the upcoming change in the [Version 2 Upgrade Guide](https://www.terraform.io/docs/providers/aws/guides/version-2-upgrade.html) and provide `[WARN]` logs indicating the change of behavior. We may also want to consider properly documenting the existing behavior and available workaround(s) in the `aws_route_table` resource documentation, even if we are deprecating it.

**Comment by YakDriver** [comment]
@bflad The `aws_route` import PR #5687 is ready to go. I'll work on cleaning up the other two PRs and update this when they are ready to go.

- [x] resource/aws_route: Support resource import #5687 (ready for review)
- [ ] resource/aws_route_table: Remove `aws_route` import #5657 (not yet ready)
- [ ] resource/aws_route_table: Document upcoming changes (no # yet) (not yet ready)
  - [ ] Document upcoming change in Version 2 Upgrade Guide
  - [ ] Provide `[WARN]` logs indicating the change in behavior
  - [ ] Document current behavior and workaround(s) in `aws_route_table` resource documentation

**Comment by YakDriver** [comment]
Acceptance tests after adjusting, rebasing.

```console
$ make testacc TESTARGS='-run=TestAccAWSRouteTable_import'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSRouteTable_import -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSRouteTable_importBasic
--- PASS: TestAccAWSRouteTable_importBasic (53.04s)
=== RUN   TestAccAWSRouteTable_importWithCreate
--- PASS: TestAccAWSRouteTable_importWithCreate (68.20s)
=== RUN   TestAccAWSRouteTable_importComplex
--- PASS: TestAccAWSRouteTable_importComplex (186.16s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	307.443s
```

**Comment by YakDriver** [comment]
@bflad Everything you suggested for implementation is ready for review, adjust, merge. Please see [earlier comment](https://github.com/terraform-providers/terraform-provider-aws/pull/5657#issuecomment-416274401) for PR #s.

**Comment by YakDriver** [comment]
This PR needs some work. Currently, with this PR:

1. **Config**: route table + inline routes
   **Import**: route table + inline routes
   **After apply**: Correct (TF reports no changes needed)
1. **Config**: route table + separate _"correctly named"_ `aws_route` resources
   **Import**: route table + inline routes + separate resources
   **After apply**: No routes deleted but 400 errors because routes already exist (because of separate resources, and none are in the state, tries to create them)
   **Expected**: Import results in route table + separate resources and TF reports no changes needed and no 400 errors because of trying to create existing resources
1. **Config**: route table + separate _incorrectly named_ `aws_route` resources
   **Import**: route table + inline routes + separate resources
   **After apply**: No routes deleted but 400 errors because routes already exist (because of separate resources, and none are in the state, tries to create them)
   **Expected**: Import results in route table + separate resources and TF reports no changes needed and no 400 errors because of trying to create existing resources

Here are configs based on these scenarios that, in my opinion, should work:

- [x] PR makes this work now.
```hcl
resource "aws_route_table" "test" {
  vpc_id = "${aws_vpc.test.id}"

  route {
    cidr_block = "0.0.0.0/0"
    gateway_id = "${aws_internet_gateway.test.id}"
  }

  tags {
    Name = "bflad-testing"
  }
}
```

- [ ] No deletion but you get 400 errors (recreating existing routes)
```hcl
resource "aws_route_table" "test" {
  vpc_id = "${aws_vpc.test.id}"

  tags {
    Name = "bflad-testing"
  }
}

resource "aws_route" "test" {
  route_table_id         = "${aws_route_table.test.id}"
  destination_cidr_block = "0.0.0.0/0"
  gateway_id             = "${aws_internet_gateway.test.id}"
}
```

- [ ] No deletion but you get 400 errors (recreating existing routes)
```hcl
resource "aws_route_table" "test" {
  vpc_id = "${aws_vpc.test.id}"

  tags {
    Name = "bflad-testing"
  }
}

resource "aws_route" "randomname" {
  route_table_id         = "${aws_route_table.test.id}"
  destination_cidr_block = "0.0.0.0/0"
  gateway_id             = "${aws_internet_gateway.test.id}"
}
```

**Comment by bflad** [comment]
#5715 has been merged so this is ready for rebase 👍 

This pull request should not worry about handling scenarios with separate `aws_route` resources in configurations -- its up to the operator to import the separate resource if they choose to manage their configuration in this manner. The new behavior of not including the other resources during import is more consistent with the import behavior of other resources in the provider. 😄

**Comment by YakDriver** [comment]
@bflad I agree with that yet there is still a problem. If you import a route table that has no routes in the configuration, you'll end up with a route table that has inline routes in the state with this PR. Shouldn't import look far enough into the configuration to know whether to include inline routes or not?

**Comment by bflad** [comment]
That's by design, not a problem.

The `route` attribute is configured as [Optional](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#optional) and [Computed](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#computed):

https://github.com/terraform-providers/terraform-provider-aws/blob/ee18327b50777d6f5e3bc06277cd723c5953ef2d/aws/resource_aws_route_table.go#L44-L48

This is what allows there to be a separate `aws_route` resource to begin with. The lack of providing a `route` configuration will show no difference in the `aws_route_table` resource. It does not matter that the routes are saved into the Terraform state as they have no bearing unless you start configuring them in your configuration.

---

As an aside: the import process for a resource is an implicit call to the read function of a resource after the `Importer.State` function is called. The import process does not currently look at or care about the configuration, only getting information into the Terraform state.

In practice, this is why we generally only care about setting the ID or required attributes for reading in the `Importer.State` functions for resources -- we allow the read function to do all the work for the rest of the Terraform state and best practices for attributes are to always re-read them into the Terraform state during the read function for drift detection: https://www.terraform.io/docs/extend/best-practices/detecting-drift.html 

While we currently support this "complex" import, where the import of one resource creates other resources in the Terraform state, its usage is very limited and often more confusing. At some point we may remove support for these "complex" imports or support better import models in coordination with upstream changes in Terraform core. There have been design sketches where Terraform automatically creates the configuration. The implementation of a feature like that may influence how import works in the future. We'll see. 😄

**Comment by YakDriver** [comment]
@bflad I guess I'll stop complaining about my PR then! Thanks

**Pull request merged**